include/lbann/data_readers/data_reader.hpp

```diff
@@ -740,13 +736,15 @@ class generic_data_reader : public lbann_image_preprocessor {
 
   lbann_comm *m_comm;
 
+  bool fetch_data_block(CPUMat& X, El::Int thread_index, El::Int mb_size, El::Matrix<El::Int>& indices_fetched, thread_pool& io_thread_pool);
+
   /**
    * Fetch a single sample into a matrix.
    * @param X The matrix to load data into.
    * @param data_id The index of the datum to fetch.
    * @param mb_idx The index within the mini-batch.
    */
-  virtual bool fetch_datum(CPUMat& X, int data_id, int mb_idx, int tid) {
+  virtual bool fetch_datum(CPUMat& X, int data_id, int mb_idx, thread_pool& io_thread_pool) {
```

Comment:
I don't think it's necessary to make the thread pool available here. This function is overridden in every derived class and they only need to know the thread index (e.g. to access per-thread buffers). It's best to concentrate complexity in the base class.

Reply:
We cannot pass the thread id into the function because we do not bind the execution context with the function when the lambda is enqueued.  As a result we need to dynamically get the thread id.

Reply:
The thread_pool is now a member variable.